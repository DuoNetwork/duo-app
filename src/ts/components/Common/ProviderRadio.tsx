//import * as d3 from 'd3';
//import moment from 'moment';
import { Wallet } from '@finbook/duo-contract-wrapper';
import { Button, Radio } from 'antd';
import * as React from 'react';
import * as CST from 'ts/common/constants';
import { web3Wrapper } from 'ts/common/wrappers';
import { SRadioGroup } from '../Cards/_styled';
import { SModal } from './_styled';

const RadioGroup = Radio.Group;
const RadioButton = Radio.Button;

interface IProps {
	refresh: () => Promise<void>;
}

interface IState {
	loading: boolean;
	accounts: string[];
	value: Wallet;
	visible: boolean;
	msg: string;
	accountIndex: number;
	showTip: boolean;
}

export default class ProviderRadio extends React.Component<IProps, IState> {
	constructor(props: IProps) {
		super(props);
		this.state = {
			loading: false,
			accounts: [],
			value: web3Wrapper.wallet,
			visible: false,
			msg: 'Please make sure Ledger is connected.',
			accountIndex: 0,
			showTip: true
		};
	}

	private handleCancel = () =>
		this.setState({
			loading: false,
			accounts: [],
			visible: false,
			msg: 'Please make sure Ledger is connected.',
			accountIndex: 0,
			showTip: true
		});

	private handleChange = (value: number) => {
		if (value === Wallet.MetaMask) {
			web3Wrapper.switchToMetaMask(window);
			this.setState({
				value: web3Wrapper.wallet
			});
			this.props.refresh();
		} else if (value === Wallet.Ledger) this.setState({ visible: true });
	};

	private handleConnect = () => {
		this.setState({ msg: 'Connecting to Leger', loading: true });
		const timer = setTimeout(() => {
			this.setState({ msg: 'Failed to connect.', loading: false });
		}, 15000);
		web3Wrapper.switchToLedger().then(accounts => {
			clearTimeout(timer);
			this.setState({
				msg: 'Select an account',
				accounts: accounts,
				loading: false,
				showTip: false
			});
		});
	};

	private handleSelect = (index: number) => this.setState({ accountIndex: index });

	private handleSubmit = async () => {
		this.setState({
			loading: true,
			value: web3Wrapper.wallet
		});
		web3Wrapper.accountIndex = this.state.accountIndex;
		await this.props.refresh();
		this.handleCancel();
	};

	public render() {
		const { accounts, loading, visible, value, msg, accountIndex, showTip } = this.state;
		return (
			<div>
				<SRadioGroup
					value={value}
					size="small"
					onChange={(e: any) => this.handleChange(e.target.value)}
				>
					<RadioButton value={Wallet.MetaMask}>{'MetaMask'}</RadioButton>
					<RadioButton value={Wallet.Ledger}>{'Ledger'}</RadioButton>
				</SRadioGroup>
				<SModal
					title="LEDGER ACCOUNT MANAGE"
					visible={visible}
					maskClosable={false}
					onCancel={this.handleCancel}
					footer={[
						<Button
							key="submit"
							type="ghost"
							loading={loading}
							onClick={accounts.length ? this.handleSubmit : this.handleConnect}
						>
							{accounts.length ? CST.TH_SUBMIT.EN : CST.TH_CONNECT}
						</Button>
					]}
				>
					{msg ? msg : null}
					<br />
					{showTip ? (
						<p className="ledger-reminder">
							<b>Contract data</b> and <b>Browser support</b> must be set to{' '}
							<b>Yes</b>
						</p>
					) : null}
					{accounts.length ? (
						<RadioGroup
							className="ledger-account-selector"
							onChange={e => this.handleSelect(e.target.value)}
							value={accountIndex}
						>
							{accounts.map((a, i) => (
								<Radio key={i} value={i}>
									{a}
								</Radio>
							))}
						</RadioGroup>
					) : null}
				</SModal>
			</div>
		);
	}
}
