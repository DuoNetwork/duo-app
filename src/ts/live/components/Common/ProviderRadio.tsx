//import * as d3 from 'd3';
//import moment from 'moment';
import { Button, Radio } from 'antd';
import * as React from 'react';
import * as CST from '../../common/constants';
import contractUtil, { Wallet } from '../../common/contractUtil';
import { SRadioGroup } from '../Cards/_styled';
import { SModal } from './_styled';

const RadioGroup = Radio.Group;
const RadioButton = Radio.Button;

interface IProps {
	refresh: () => Promise<void>;
}

interface IState {
	loading: boolean;
	accounts: string[];
	value: Wallet;
	visible: boolean;
	msg: string;
	accountIndex: number;
	showTip: boolean;
}

export default class ProviderRadio extends React.Component<IProps, IState> {
	constructor(props: IProps) {
		super(props);
		this.state = {
			loading: false,
			accounts: [],
			value: contractUtil.wallet,
			visible: false,
			msg: 'Please make sure Ledger is connected.',
			accountIndex: 0,
			showTip: true
		};
	}

	private handleCancel = () =>
		this.setState({
			loading: false,
			accounts: [],
			visible: false,
			msg: 'Please make sure Ledger is connected.',
			accountIndex: 0,
			showTip: true
		});

	private handleChange = (value: number) => {
		if (value === Wallet.MetaMask) {
			contractUtil.switchToMetaMask();
			this.setState({
				value: contractUtil.wallet
			});
			this.props.refresh();
		} else if (value === Wallet.Ledger) this.setState({ visible: true });
	};

	private handleConnect = () => {
		this.setState({ msg: 'Connecting to Leger', loading: true });
		const timer = setTimeout(() => {
			this.setState({ msg: 'Failed to connect.', loading: false });
		}, 15000);
		contractUtil.switchToLedger().then(accounts => {
			clearTimeout(timer);
			this.setState({
				msg: 'Select an account',
				accounts: accounts,
				loading: false,
				showTip: false
			});
		});
	};

	private handleSelect = (index: number) => this.setState({ accountIndex: index });

	private handleSubmit = async () => {
		this.setState({
			loading: true,
			value: contractUtil.wallet
		});
		contractUtil.accountIndex = this.state.accountIndex;
		await this.props.refresh();
		this.handleCancel();
	};

	public render() {
		const { accounts, loading, visible, value, msg, accountIndex, showTip } = this.state;
		return (
			<div>
				<SRadioGroup
					value={value}
					size="small"
					onChange={e => this.handleChange(e.target.value)}
				>
					<RadioButton value={Wallet.MetaMask}>{'MetaMask'}</RadioButton>
					<RadioButton value={Wallet.Ledger}>{'Ledger'}</RadioButton>
				</SRadioGroup>
				<SModal
					title="LEDGER ACCOUNT MANAGE"
					visible={visible}
					maskClosable={false}
					onCancel={this.handleCancel}
					footer={[
						<Button
							key="submit"
							type="ghost"
							loading={loading}
							onClick={accounts.length ? this.handleSubmit : this.handleConnect}
						>
							{accounts.length ? CST.TH_SUBMIT : CST.TH_CONNECT}
						</Button>
					]}
				>
					{msg ? msg : null}
					{showTip ? (
						<p className="ledger-reminder">
							<b>Contract data</b> and <b>Browswe support</b> must be set to{' '}
							<b>Yes</b>
						</p>
					) : null}
					{accounts.length ? (
						<RadioGroup
							onChange={e => this.handleSelect(e.target.value)}
							value={accountIndex}
						>
							{accounts.map((a, i) => (
								<Radio key={i} value={i}>
									{a}
								</Radio>
							))}
						</RadioGroup>
					) : null}
				</SModal>
			</div>
		);
	}
}
