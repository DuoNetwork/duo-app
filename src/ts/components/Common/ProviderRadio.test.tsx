import { shallow } from 'enzyme';
import * as React from 'react';
// import { IConversion } from 'ts/common/types';
import ProviderRadio from './ProviderRadio';

describe('AuthCard Test', () => {
	describe('User Login', () => {

		const refresh = jest.fn();

		it('test Snapshot', () => {
			const wrapper = shallow(<ProviderRadio refresh={refresh} />);
			expect(wrapper).toMatchSnapshot();
		});
	});
});
