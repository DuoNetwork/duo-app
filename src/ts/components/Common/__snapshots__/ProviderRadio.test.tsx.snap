// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProviderRadio Test User Login test Snapshot 1`] = `
<div>
  <Styled(RadioGroup)
    onChange={[Function]}
    size="small"
    value={0}
  >
    <RadioButton
      value={2}
    >
      MetaMask
    </RadioButton>
    <RadioButton
      value={3}
    >
      Ledger
    </RadioButton>
  </Styled(RadioGroup)>
  <Styled(Modal)
    footer={
      Array [
        <Button
          block={false}
          ghost={false}
          htmlType="button"
          loading={false}
          onClick={[Function]}
          type="ghost"
        >
          Connect
        </Button>,
      ]
    }
    maskClosable={false}
    onCancel={[Function]}
    title="LEDGER ACCOUNT MANAGE"
    visible={false}
  >
    Please make sure Ledger is connected.
    <br />
    <p
      className="ledger-reminder"
    >
      <b>
        Contract data
      </b>
       and 
      <b>
        Browser support
      </b>
       must be set to
       
      <b>
        Yes
      </b>
    </p>
  </Styled(Modal)>
</div>
`;

exports[`ProviderRadio Test User Login test Snapshot 2`] = `
<div>
  <Styled(RadioGroup)
    onChange={[Function]}
    size="small"
    value={0}
  >
    <RadioButton
      value={2}
    >
      MetaMask
    </RadioButton>
    <RadioButton
      value={3}
    >
      Ledger
    </RadioButton>
  </Styled(RadioGroup)>
  <Styled(Modal)
    footer={
      Array [
        <Button
          block={false}
          ghost={false}
          htmlType="button"
          loading={false}
          onClick={[Function]}
          type="ghost"
        >
          Connect
        </Button>,
      ]
    }
    maskClosable={false}
    onCancel={[Function]}
    title="LEDGER ACCOUNT MANAGE"
    visible={false}
  >
    Please make sure Ledger is connected.
    <br />
    <p
      className="ledger-reminder"
    >
      <b>
        Contract data
      </b>
       and 
      <b>
        Browser support
      </b>
       must be set to
       
      <b>
        Yes
      </b>
    </p>
  </Styled(Modal)>
</div>
`;

exports[`ProviderRadio Test User Login test Snapshot 3`] = `
<div>
  <Styled(RadioGroup)
    onChange={[Function]}
    size="small"
    value={0}
  >
    <RadioButton
      value={2}
    >
      MetaMask
    </RadioButton>
    <RadioButton
      value={3}
    >
      Ledger
    </RadioButton>
  </Styled(RadioGroup)>
  <Styled(Modal)
    footer={
      Array [
        <Button
          block={false}
          ghost={false}
          htmlType="button"
          loading={false}
          onClick={[Function]}
          type="ghost"
        >
          Connect
        </Button>,
      ]
    }
    maskClosable={false}
    onCancel={[Function]}
    title="LEDGER ACCOUNT MANAGE"
    visible={true}
  >
    Please make sure Ledger is connected.
    <br />
    <p
      className="ledger-reminder"
    >
      <b>
        Contract data
      </b>
       and 
      <b>
        Browser support
      </b>
       must be set to
       
      <b>
        Yes
      </b>
    </p>
  </Styled(Modal)>
</div>
`;
